iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperations.java

```diff
@@ -282,6 +311,98 @@ protected void doCommit(TableMetadata base, TableMetadata metadata) {
     }
   }
 
+  /**
+   * Build a new partition spec with new schema from original pspec. The new pspec has the same
+   * partition fields as the original pspec with source ids from the new schema
+   *
+   * @param originalPspec
+   * @param originalSchema
+   * @param newSchema
+   * @return new partition spec
+   */
+  public static PartitionSpec rebuildPartitionSpec(
+      PartitionSpec originalPspec, Schema originalSchema, Schema newSchema) {
+    PartitionSpec.Builder builder = PartitionSpec.builderFor(newSchema);
+
+    for (PartitionField field : originalPspec.fields()) {
+      // get field name from original schema using source id of partition field
+      // because Pspec appends _bucket and _trunc to field name for bucket and truncate fields
+      String fieldName = originalSchema.findField(field.sourceId()).name();
+      // Check if the partition field is present in new schema
+      if (newSchema.findField(fieldName) == null) {
+        throw new IllegalArgumentException(
+            "Field " + fieldName + " does not exist in the new schema");
+      }
+
+      // Recreate the transform using the string representation
+      String transformString = field.transform().toString();
+
+      // Add the field to the new PartitionSpec based on the transform type
+      if ("identity".equalsIgnoreCase(transformString)) {
+        builder.identity(fieldName);
+      } else if (transformString.startsWith("bucket[")) {
+        // Extract bucket number from the string (e.g., bucket[16])
+        int numBuckets =
+            Integer.parseInt(
+                transformString.substring(
+                    transformString.indexOf('[') + 1, transformString.indexOf(']')));
+        builder.bucket(fieldName, numBuckets);
+      } else if (transformString.startsWith("truncate[")) {
+        // Extract width from the string (e.g., truncate[10])
+        int width =
+            Integer.parseInt(
+                transformString.substring(
+                    transformString.indexOf('[') + 1, transformString.indexOf(']')));
+        builder.truncate(fieldName, width);
+      } else if ("year".equalsIgnoreCase(transformString)) {
+        builder.year(fieldName);
+      } else if ("month".equalsIgnoreCase(transformString)) {
+        builder.month(fieldName);
+      } else if ("day".equalsIgnoreCase(transformString)) {
+        builder.day(fieldName);
+      } else if ("hour".equalsIgnoreCase(transformString)) {
+        builder.hour(fieldName);
+      } else {
+        throw new UnsupportedOperationException("Unsupported transform: " + transformString);
+      }
+    }
```

Comment:
this should be placed in a separate method where the transform is reconstructed.

Reply:
Separate method is not such a good idea here because we will be mutating the pspec builder in there and then build the pspec outside of it.

Reply:
not following. what's the problem of creating a method only having side effects

Reply:
done